### DISCUSSION TITLE
Timeline for Mill 0.13.0

### DISCUSSION THREAD

**Discussion by lihaoyi** [com-lihaoyi/mill#3901] [Development]
This is a thread to centralize discussion about the plans for 0.13.0, now that 0.12.0 has landed. A follow up to https://github.com/com-lihaoyi/mill/discussions/3209

## What is Mill 0.13.0

0.13.0 will be a binary incompatible breaking release for Mill. We accumulated a backlog of issues that could not go into 0.12.0 due to binary compatibility breakage, which will need to go into 0.13.0:

- Issues https://github.com/com-lihaoyi/mill/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.13.0
- Pull Requests https://github.com/com-lihaoyi/mill/pulls?q=is%3Aopen+is%3Apr+milestone%3A0.13.0

The Scala 3 upgrade PR (https://github.com/com-lihaoyi/mill/pull/3369) is probably the first such PR that we will need to bring up to date and merge, but there's a significant backlog of TODO issues and PRs, and I expect the removal of binary compatibility constraints will give us the opportunity for further cleanups and improvements that we don't know about yet.

## When is Mill 0.13.0 Happening?

For now my proposed timeline for 0.13.0 is as follows:

1. Oct 2024: Mill 0.12.0 final is released (DONE)
    * `main` still targets `0.12.x` series, which will need to be (eventually) merged into `0.13.0` PRs (primarily https://github.com/com-lihaoyi/mill/pull/3369)
2. Jan 2025: `0.12.x` maintenance branch is cut, first binary-incompatible changes land in `main` targeting `0.13.0`
    * `main` now targets `0.13.0`, PRs which want to target `0.12.x` will need to be explicitly backported to `0.12.x` branch
4. Apr 2025: `0.13.0` RCs begin to go out
5. May 2025: `0.13.0` final is released, third-party plugins are re-published

This is a reasonably long timeline, but `0.13.0` is our chance to make binary-incompatible changes. 

* We want to wait until the `0.12.x` series has stabilized before making `main` target `0.13.0`, since after that we would need to backport fixes to the `0.12.x` maintenance branch, which is annoying

* The more such changes we can batch together and the more we can get them correct, the longer we can wait for the next breaking release of 0.14.0, which is ultimately good for the stability of the ecosystem. So we want to take our time preparing 0.13.0

* At the same time, we do not want the "PRs targeting main/0.13.0 cannot be released since 0.13.0 isn't ready yet" phase to last too long, since that means any merged PRs that people want to actually used would need to be backported to `0.12.x`

Hence the 2-3 month period from Oct-Jan to stabilize `0.12.x` and the 3-4 month period from Jan to Apr/May to prepare `0.13.0` is a compromise between these three goals.

Hopefully, if we do this right, the outcome will be that we'll be able continue active development of Mill with new features and improvements, while also preserving binary/semantic compatibility in the `0.13.x` series as long as possible.

**Comment by lihaoyi** [comment]
We are now a good way through our Mill 0.12.x cycle, with 0.12.3 released. I expect to have maybe 2 more 0.12.x releases - 0.12.4 in mid-dec 2024, 0.12.5 in mid-jan 2025 - before merging https://github.com/com-lihaoyi/mill/pull/3369 and breaking compatibility in `main` and cutting a `0.12.x` release branch. 

After that, PRs to `0.12.x` can continue, but will be less convenient since they will require backporting with duplicate PRs targeting `main` and `0.12.x`

**Comment by lihaoyi** [comment]
Mill 0.12.5 is out Mill 0.12.5 is out https://github.com/com-lihaoyi/mill#0-12-5, with a lot of polish and small improvements

Probably one more 0.12.6 release in mid-january before we break bincompat and start working towards 0.13.0

**Comment by lihaoyi** [comment]
0.12.6 is due this/next week. The last major group of changes I'd like to get into 0.12.6 is @alexarchambault's:

- https://github.com/com-lihaoyi/mill/discussions/4281. 

These look like they're mostly done, and just need to be reviewed and merged.

The Mill native launcher has landed in master, and is available for people to try using the latest unstable release (`0.12.5-68-e4bf78`) and using the latest `mill`/`mill.bat` bootstrap script in the Mill repo.

**Comment by lihaoyi** [comment]
Looks like https://github.com/com-lihaoyi/mill/pull/4363 and https://github.com/com-lihaoyi/mill/issues/4353 have appeared as a new PR to merge and new issue to investigate and resolve before we can cut 0.12.6. May push back the release by a week or so but should get it out before end of the month

**Reply by lefou** [reply]
Another blocker:
* https://github.com/com-lihaoyi/mill/issues/4378

**Comment by lihaoyi** [comment]
Looks like all the big PRs have merged; I'm aiming to cut 0.12.6 on Friday 31 January 2025. The next few days until then will mostly just be housekeeping and documentation work

After 0.12.6 is cut, I will update and merge the Scala 3 upgrade into the main branch the next day.

**Reply by lefou** [reply]
Can we please tag a `0.13.0-M0` or `0.13.0-RC0` before merging anything incompatible to `0.12.x`. That results in snapshot versions properly comunicating their target.

**Reply by lihaoyi** [reply]
@lefou shouldn't we tag 0.13.0-M0 _after_ merging the first incompatible PR? Since anything before the first incompatible PR is still compatible with 0.12.x

**Reply by lefou** [reply]
No, since we derive the version from the last tag, the first incompatible merge would end up with a version that suggests compatibility.

**Reply by lihaoyi** [reply]
Sure, but  if we tag `0.13.0-M0` before doing the first incompatible merge, that means that `0.13.0-M0` will be the last tag on both `main` as well as `0.12.x`, so any unstable releases we cut from `0.12.x` will be of the form `0.13.0-M0-blah-blah`. 

If we tag `0.13.0-M0` after doing the first incompatible merge, then the `0.13.0-M0` tag will only appear on the history of `main` and not of `0.12.x`; that means that any unstable releases cut from `main` will be `0.13.0-M0-blah-blah`, and any unstable releases cut from `0.12.x` will be of `0.12.7-blah-blah`. This seems more correct

**Reply by lefou** [reply]
Let the first commit only change non-code stuff, e.g. disable mandatory mima checks and tag it `0.13.0-M0`. Then branch of a `0.12.x` one commit before.

**Comment by lihaoyi** [comment]
0.12.6 is released https://github.com/com-lihaoyi/mill/blob/main/changelog.adoc#0126---2025-01-31

**Comment by lihaoyi** [comment]
Prepared the `0.12.x`, merged the first incompatible stuff into `main`, and tagged `0.13.0-M0` on he first merge. Ended up just merging https://github.com/com-lihaoyi/mill/pull/3369 whole.

**Reply by lefou** [reply]
I think c05e81bc0287895e39accfdeec65f8f49b0e039c would be the best commit to start a `0.12.x` branch of.

**Reply by lefou** [reply]
Turns out, the `0.12.x` branch already existed but then got accidentally deleted when PR https://github.com/com-lihaoyi/mill/pull/4454 was merged. I restored it now.
